### PR TITLE
[web] replace browser-related conditional logic with BrowserEnvironment

### PR DESCRIPTION
**Pull request by yjbanov**
In order to deal with browser differences we have relied on conditional logic (`if` blocks) and map look-ups scattered throughout the felt codebase. For example, there would be chunks of code like this:

```dart
    if (isSafariIOS) {
      await IosSafariArgParser.instance.initIosSimulator();
    }
```

As well as listings and mapping like [this](https://github.com/flutter/engine/blob/6c2b052383c6b2964bbb756ea05500a919946769/lib/web_ui/dev/supported_browsers.dart#L27).

This PR introduces `BrowserEnvironment` interface that provides the functionality needed by the test runner. A single `switch` block at the root (in `test_runner.dart`) is used to determine a concrete implementation of this interface. The rest of the code no longer needs to deal with browser differences. It just requests the needed functionality from `BrowserEnvironment`.

As a result all the routing logic from `supported_browsers.dart` is gone and the file deleted. Browser-specific implementations of `ScreenshotManager` are moved to their respective browser files (`chrome.dart`, `firefox.dart`, etc), and the `screenshot_manager.dart` file deleted.

The logic in `safari.dart` that attempted to support desktop and mobile Safari vis `if`/`else` blocks has been split up into `safari_ios.dart` and `safari_macos.dart`.

The rest of the code was updated to use the `BrowserEnvironment` API.

Bonus: `do-fetch-goldens-repo` has been changed to `skip-goldens-repo-fetch` to better reflect the default (you opt-in to skip the fetch rather than ask it to fetch).

### CODE REVIEW

**Review by mdebbar — APPROVED** [review]
I love this new structure. Makes much sense!

File: lib/web_ui/dev/safari_ios.dart [hunk 1]
```diff
@@ -2,89 +2,86 @@
...
+  ScreenshotManager? getScreenshotManager() {
+    return SafariIosScreenshotManager();
+  }
+
+  @override
+  String get packageTestConfigurationYamlFile => 'dart_test_safari.yaml';
+}
 
-    return screenshot;
+/// A class for running an instance of Safari for iOS (i.e. mobile Safari).
```
[the remark this answers is not in the document]
Comment (yjbanov) [reply lib/web_ui/dev/safari_ios.dart:45]: Thanks for rereading the comments. I didn't read the comments for code that I copied (although I should have). Fixed here and a few other places.

File: lib/web_ui/dev/browser.dart [hunk 2]
```diff
@@ -147,3 +186,18 @@ abstract class Browser {
...
+  /// Capture a screenshot.
+  ///
+  /// Please read the details for the implementing classes.
+  Future<Image> capture(math.Rectangle region);
+
+  /// Suffix to be added to the end of the filename.
+  ///
+  /// Example file names:
+  /// - Chrome, no-suffix: backdrop_filter_clip_moved.actual.png
+  /// - iOS Safari: backdrop_filter_clip_moved.actual.iOS_Safari.png
```
[the remark this answers is not in the document]
Comment (yjbanov) [reply lib/web_ui/dev/browser.dart:201]: Done.

File: lib/web_ui/dev/safari_ios.dart [hunk 3]
```diff
@@ -2,89 +2,86 @@
...
+    return SafariIos._(() async {
+      // iOS-Safari
+      // Uses `xcrun simctl`. It is a command line utility to control the
+      // Simulator. For more details on interacting with the simulator:
+      // https://developer.apple.com/library/archive/documentation/IDEs/Conceptual/iOS_Simulator_Guide/InteractingwiththeiOSSimulator/InteractingwiththeiOSSimulator.html
+      var process = await io.Process.start('xcrun', [
+        'simctl',
+        'openurl', // Opens the url on Safari installed on the simulator.
+        'booted', // The simulator is already booted.
+        '${url.toString()}'
```
[the remark this answers is not in the document]
Comment (yjbanov) [reply lib/web_ui/dev/safari_ios.dart:68]: Done.

**Pull request merged**